Improve docs on scope and target platforms. (#92)
unicode-org/icu4x: docs/charter.md

ICU4X Charter
==============

The ICU4X project designs and develops a set of modular internationalization components suitable for use in client-side and resource-constrained environments.

ICU4X builds on the design and API decisions of [ICU4C]/[ICU4J] and [ECMA-402].

## Design

ICU4X is being built from the start with several key design constraints:

1. Small and modular code.
2. Pluggable locale data.
3. Availability and ease of use in multiple programming languages.
4. Written by i18n experts to encourage best practices.

## Scope

The scope of ICU4X is the set of functionality covered by [ECMA-402], with extensions to cover the needs of clients including Gecko and Fuchsia that are consistent with the spirit of [ECMA-402].

APIs included in ICU4X will:

* Be useful to a broad audience of client-side applications
* Depend on Unicode or CLDR data or algorithms

> * Benefits from being versioned alongside Unicode and CLDR data

Reviewer: Benefit vs Benefits

ICU4X Charter
==============

> The ICU4X project designs and develops a set of modular internationalization components suitable for use on the client-side and in resource-constrained environments.

Reviewer 2: "client-side", with a hyphen, is an adjective.  You should say either "for use on the client side and in resource-constrained environments" or "for use in client-side or resource-constrained environments".

(I'm being extra picky since this is the first sentence of the charter)

> ICU4X builds on the design and API decisions of [`ICU4C`]/[`ICU4J`] and [`ECMA-402`].

Reviewer 2: Optional: Using backticks seems like an odd style choice.  I would just write `[ICU4C]` (and ICU4J, ECMA-402).  They will turn to blue links, which I think is enough to highlight them from the background.

## Design

ICU4X is being built from the start with several key design constraints:

1. Small and modular code.
2. Pluggable locale data.
3. Availability and ease of use in multiple programming languages.
4. Written by i18n experts to encourage best practices.

## Scope

The scope of ICU4X is the set of functionality covered by [`ECMA-402`], with extensions to cover the needs of clients including Gecko and Fuchsia that are consistent with the spirit of [`ECMA-402`].

> One of the benchmarks on decisions related to whether a given API belongs to ICU4X will be if it requires Unicode/CLDR data and would benefit from being managed together with other APIs that rely on the same version of Unicode and CLDR data.

Reviewer 2: Suggestion: This paragraph seems redundant with the third bullet point.  I would probably remove it.

In some rare cases, ICU4X may diverge from both [`ECMA-402`], and [`ICU4C`]/[`ICU4J`]. Such diversions should be the result of improved API design and should allow users to cover the same needs with the upgraded APIs.

## Target platforms

The list of target platforms is expected to evolve over time, and will be primarily used when making decisions on features and API tradeoffs between low-level ICU4X-like APIs and high-level [`ECMA-402`] APIs.

Current list of target platforms:
* Web Platform (V8, SpiderMonkey, JSC)
* Software platforms (Fuschia, Gecko)
* Mobile OSes (iOS, Android)
* Low-power OSes with [`alloc`] (WearOS, WatchOS)
* Client-side toolkits (Flutter)

ICU4X is also aiming to provide bindings or compilation targets for a range of programming languages. This list is also dynamic and used when evaluating design tradeoffs:

* Rust
* JavaScript
* Objective-C
* Java
* Dart
* C++

Referential platforms and languages may change over time as the industry developers and should represent the current needs of the industry.

> A viable subset of ICU4X will be targeting the [`no_std`] support and in the future we may explore the [`no_std`]+[`alloc`] compatibility for viable components.

Reviewer 2: Suggestion/Optional: Link to https://github.com/unicode-org/icu4x/issues/77

ICU4X Charter
==============

> ICU4X is a project whose objective is to solve the needs of clients who wish to provide client-side i18n for their products in resource-constrained environments.

Reviewer 3: "objective to solve the needs…" is perhaps too many levels of indirection. How about something like, "The ICU4X project provides client-side i18n for resource-constrainied environments"?

> ICU4X derives from the experience of ICU4C/ICU4J and ECMA402 design decisions.

Reviewer 3: ICU4X builds on the design and API decisions of ICU4C/ICU4X and ECMA-402?

(should use consistent naming for ECMA-402 …)

Also maybe endnote links to ICU[CJ], ECMA, etc…

3. Availability and ease of use in multiple programming languages.
4. Written by i18n experts to encourage best practices.

# Scope

> ICU4X aims to provide a set of APIs sufficient to internationalize basic modern software.

Reviewer 3: "aim" and "objective" seem somewhat speculative, I know this is still early, but shouldn't this be written with the 'final' product in mind?

> With the focus on client-side, it will initially focus on the scope covered by the ECMA-402 extended to cover needs of more low level systems such as Gecko or Fuchsia.

Reviewer 3: ```suggestion
With the client-side focus, the initial scope is that covered by ECMA-402, with some extensions to cover needs of low level systems such as Gecko or Fuchsia.
```

Reviewer 2: My version:

> The scope of ICU4X is the set of functionality covered by ECMA-402, with extensions to cover the needs of clients including Gecko and Fuchsia that are consistent with the spirit of ECMA-402.

The current target is to cover all of ECMA-402 needs, and a modern subset of ICU4X APIs needed for software platforms such as Fuschia and Gecko.

> One of the benchmarks on decisions related to whether a given API belongs to ICU4X will be if it requires Unicode/CLDR data and would benefit from being managed together with other APIs that rely on the same version of Unicode and CLDR data.

Reviewer 2: This is the most important paragraph in my opinion.  We should list out what all of the benchmarks are.  Suggestions:

- Useful to a broad audience of client-side applications
- Depends on Unicode or CLDR data or algorithms
- Benefits from being versioned alongside Unicode and CLDR data

Reviewer 3: i have a suggested change below 👇

> The current target is to cover all of ECMA-402 needs, and a modern subset of ICU4X APIs needed for software platforms such as Fuschia and Gecko.
> 
> One of the benchmarks on decisions related to whether a given API belongs to ICU4X will be if it requires Unicode/CLDR data and would benefit from being managed together with other APIs that rely on the same version of Unicode and CLDR data.
> 
> For ICU4C APIs and other internationalization needs that will not end up being included, ICU4X aims to provide a sufficient access to building blocks to allow third-party libraries to be developed that can interoperate with ICU4X with minimum data and logic duplication.
> 
> In some rare cases, ICU4X may diverge from both ECMA-402, and ICU4C. Such diversions should be the result of improved API design and should allow users to cover the same needs with the upgraded APIs.

Reviewer 3: first paragraph duplicates previous one

```suggestion
Two of the benchmarks for inclusion in ICU4X are: (1) if the API requires Unicode/CLDR data and (2) if it would benefit from being managed with other APIs that rely on the same version of Unicode and CLDR data.

For such APIs and functionality that will not end up being included, ICU4X aims to provide the building blocks to enable third-party libraries to be developed which depend on and interoperate with ICU4X, with minimum data and logic duplication.

In some rare cases, ICU4X may diverge from both ECMA-402, and ICU4C/ICU4J. Such diversions should be the result of improved API design and should allow users to cover the same needs with the upgraded APIs.
```

# Target platforms

The list of target platforms is expected to evolve over time, and will be primarily used when making decisions on features and API tradeoffs between low-level ICU4X-like APIs and high-level ECMA-402 APIs.

Current list of target platforms:
* Web Platform (V8, SpiderMonkey, JSC)
* Software platforms (Fuschia, Gecko)
* Mobile OSes (iOS, Android)
* Low-power OSes with `alloc` (WearOS, WatchOS)
* Client-side toolkits (Flutter)

ICU4X is also aiming to provide bindings or compilation targets for a range of programming languages. This list is also dynamic and used when evaluating design tradeoffs:

* Rust
* JavaScript
* Objective-C
* Java
* Dart
* C++

Referential platforms and languages may change over time as the industry developers and should represent the current needs of the industry.

> A viable subset of ICU4X will be targeting the `no_std` support and in the future we may explore the `no_std+alloc` compatibility for viable components.

Reviewer 3: (Maybe links to `no_std` / `no_std+alloc` documentation ? )

### Is ICU4X going to replace ICU?

> No!

Reviewer 2: Why did you delete this line? Please add it back unless you think people need to read the full paragraph in order to answer this question.

Author: Because it doesn't add a lot of value and doesn't feel like an official document.
A paragraph is not long so I think a more neuanced answer should be ok. 

I can bring it back, if you are strongly in favor of keeping it. Just sync with @nciric on how it should look like.

Reviewer 2: OK, if @zbraniecki and @nciric think the line is not necessary, let's remove it.  I just wanted to make sure it wasn't removed accidentally.

> No!

Reviewer: No! what?

Reviewer 2: No to answer the question right above it.

### How will you actually implement ICU4X?

> Internally, the first draft of ICU4X will be built in a [`no_std`], with plans to use either FFI or [WebAssembly](https://webassembly.org/) for porting the code to the other target platforms; however, we are leaving the door open for other options such as transpilation that can produce code directly.  For example, we expect Rust clang/llvm tools to stabilize in the future, which could be the basis of a potential transpilation effort.  We are also looking at a Lisp-like source that can generate high-level code.

Reviewer 2: `"in a [no_std] Rust environment"` (you lost the "Rust environment")
